pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -98,88 +105,218 @@ func (r *MachineSetSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	}
+
+	// Convert the CAPI MachineSet and AWS resources to a MAPI MachineSet
+	convertedMachineSet, warns, err := capi2mapi.FromMachineSetAndAWSMachineTemplateAndAWSCluster(
+		capiMachineSet, template, cluster,
+	).ToMachineSet()
+
+	if err != nil {
+		logger.Error(err, "Failed to convert CAPI MachineSet to MAPI machine set")
+
+		r.updateSynchronizedConditionWithPatch(
+			ctx, mapiMachineSet, corev1.ConditionFalse,
+			"FailedToConvertMachineSet", err.Error(),
+		)
+
+		return ctrl.Result{}, fmt.Errorf("failed to convert CAPI MachineSet to MAPI machine set: %w", err)
+	}
+
+	for _, warning := range warns {
+		logger.Info("Warning during conversion", "warning", warning)
+		r.Recorder.Event(mapiMachineSet, corev1.EventTypeWarning, "ConversionWarning", warning)
+	}
+
+	convertedMachineSet.Spec.Template.Labels = util.MergeMaps(mapiMachineSet.Spec.Template.Labels, convertedMachineSet.Spec.Template.Labels)
+
+	// Check if there are any changes after updating the spec
+	if !reflect.DeepEqual(convertedMachineSet.Spec, mapiMachineSet.Spec) {
+		logger.Info("Updating MAPI machine set spec")
+
+		if err := r.Update(ctx, convertedMachineSet); err != nil {
+			logger.Error(err, "Failed to update MAPI machine set")
+
+			r.updateSynchronizedConditionWithPatch(
+				ctx, mapiMachineSet, corev1.ConditionFalse,
+				"FailedToUpdateMAPI", err.Error(),
+			)
+
+			return ctrl.Result{}, fmt.Errorf("failed to update MAPI machine set: %w", err)
+		}
+
+		logger.Info("Successfully updated MAPI machine set spec")
+	} else {
+		logger.Info("No changes detected in MAPI machine set spec")
+	}
+
+	r.updateSynchronizedConditionWithPatch(ctx, convertedMachineSet, corev1.ConditionTrue,
+		consts.ReasonResourceSynchronized, "Synchronized CAPI to MAPI", capiMachineSet.Generation)
+
 	return ctrl.Result{}, nil
 }
 
+func (r *MachineSetSyncReconciler) updateSynchronizedConditionWithPatch(ctx context.Context, mapiMachineSet *machinev1beta1.MachineSet, status corev1.ConditionStatus, reason, message string, generation ...int64) {
+	logger := log.FromContext(ctx).WithValues("function", "updateSynchronizedConditionWithPatch")
+	machinesetCopy := mapiMachineSet.DeepCopy()
+
+	var newCondition *machinev1beta1.Condition
+	if status == corev1.ConditionTrue {
+		newCondition = conditions.TrueConditionWithReason(
+			consts.SynchronizedCondition, reason, message, //nolint:govet
+		)
+
+		// Update the synchronizedGeneration
+		mapiMachineSet.Status.SynchronizedGeneration = generation[0]
+	} else {
+		newCondition = conditions.FalseCondition(
+			consts.SynchronizedCondition, reason,
+			machinev1beta1.ConditionSeverityError, message, //nolint:govet
+		)
+	}
+
+	conditions.Set(&mapiMachineSet.Status.Conditions, newCondition)
+
+	// Create a Patch object using Apply
+	patch := client.MergeFrom(machinesetCopy)
+
+	data, err := patch.Data(mapiMachineSet)
+	if err != nil {
+		logger.Error(err, "failed serialising patch for log")
+	}
+	logger.Info("contents of patch", "data", data)
```

Comment:
Do we always want to log this?

Reply:
Oop - I"m still working out SSA Patching, this is a debug line :P